Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp

```diff
@@ -171,63 +468,109 @@ void Client::LookForWorkers()
         return;
     }
 
-    // randomize the start index to better distribute workers when there
+    // find someone to connect to
+    // randomize the start indices to better distribute workers when there
     // are many workers/clients - otherwise all clients will attempt to connect
     // to the same subset of workers
-    Random r;
-    size_t startIndex = r.GetRandIndex( (uint32_t)numWorkers );
-
-    // find someone to connect to
-    for ( size_t j=0; j<numWorkers; j++ )
+    // get a random worker
+    Random randomWorker;
+    size_t workerStartIndex = randomWorker.GetRandIndex( (uint32_t)numWorkers );
+    Array< AString > newlyExcludedWorkers;
+    for ( size_t i=0; i<numWorkers; ++i )
     {
-        const size_t i( ( j + startIndex ) % numWorkers );
+        const size_t j( ( i + workerStartIndex ) % numWorkers );
+        const AString & worker = m_WorkerList[ j ];
+        ServerState * ss = nullptr;
+        for ( size_t m=0; m<numWorkers; ++m )
+        {
+            ServerState * const stateToCheck = m_ServerList[ m ];
+            if ( stateToCheck && stateToCheck->m_RemoteName == worker )
```

Comment:
It should be impossible for stateToCheck to be nullptr.  All other iteration of m_ServerList asserts that the pointers are non-nullptr.